builtin/logical/aws/secret_access_keys.go

```diff
@@ -43,26 +47,45 @@ func secretAccessKeys(b *backend) *framework.Secret {
 	}
 }
 
-func genUsername(displayName, policyName, userType string) (ret string, warning string) {
-	var midString string
+func genUsername(displayName, policyName, userType, usernameTemplate string) (ret string, err error) {
 
 	switch userType {
 	case "iam_user":
 		// IAM users are capped at 64 chars; this leaves, after the beginning and
 		// end added below, 42 chars to play with.
-		midString = fmt.Sprintf("%s-%s-",
-			normalizeDisplayName(displayName),
-			normalizeDisplayName(policyName))
-		if len(midString) > 42 {
-			midString = midString[0:42]
-			warning = "the calling token display name/IAM policy name were truncated to fit into IAM username length limits"
+		up, err := template.NewTemplate(template.Template(usernameTemplate))
+		if err != nil {
+			return "", fmt.Errorf("unable to initialize username template: %w", err)
+		}
+
+		um := UsernameMetadata{
+			DisplayName: normalizeDisplayName(displayName),
+			PolicyName:  normalizeDisplayName(policyName),
+		}
+
+		ret, err = up.Generate(um)
+		if err != nil {
+			return "", fmt.Errorf("failed to generate username: %w", err)
+		}
+		// To prevent template from exceeding IAM length limits
+		if len(ret) > 64 {
+			ret = ret[0:64]
+			// warning = "the calling token display name/IAM policy name were truncated to 64 characters to fit within IAM username length limits"
```

Comment:
@calvn I updated `genUsername` to return an error instead of warnings. Now all functions that call `genUsername` return `logical.ErrorResponse` objects if they get an error. I was wondering how to now display this truncating warning to the user. Should I return a warning object as well and display the warnings if there are any?

Reply:
Returning `logical.ErrorResponse` translates to a 400 error code response, but it seems like this is more of an internal-type error, so we should return it as a regular error. Vault will return a 500 along with the error if returned this way.

The warning actually highlights a good point -- we used to truncate the policy and display name portion of the string, which means that `(unix_time) (random 20)` on the end always ensures uniqueness. Since we're now simply truncating at the end, this could lead to non-unique usernames to be returned if the policy or display name is long enough.

Reply:
This should work, but you might want to add a test against it:
`{{ printf "vault-%s-%s-%s" (printf "%s-%s" (.DisplayName) (.PolicyName) | truncate 42) (unix_time) (random 20) | truncate 64 }}`

You'd also have to re-introduce the logic to check and return the warning. You could return `(ret string, warning string, err error)`.

Reply:
@calvn This would ensure that all usernames generated by the default template above will be less than 64 chars, but how do we ensure that the template that the user provides will also generate unique usernames that are under 64 chars?

Reply:
You'll likely have to keep the `if len(ret) > 64 {...}` check to ensure that custom templates also don't exceed this limit.

Reply:
> You'd also have to re-introduce the logic to check and return the warning. You could return (ret string, warning string, err error).

Now that users can specify custom templates, the 42-length truncation warning might not make sense if they have some other non-default template. I'm also okay with removing it and moving it into the docs for visibility.

The 64 length truncation might be worth returning though.

Reply:
That makes sense! Added the refactors :)